src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -88,16 +96,54 @@ void MulticopterLandDetector::updateSubscriptions()
 	orb_update(ORB_ID(vehicle_status), _vehicleStatusSub, &_vehicleStatus);
 	orb_update(ORB_ID_VEHICLE_ATTITUDE_CONTROLS, _actuatorsSub, &_actuators);
 	orb_update(ORB_ID(actuator_armed), _armingSub, &_arming);
+	orb_update(ORB_ID(sensor_combined), _sensorsSub, &_sensors);
 }
 
-bool MulticopterLandDetector::update()
+LandDetectionResult MulticopterLandDetector::update()
 {
 	// first poll for new data from our subscriptions
 	updateSubscriptions();
 
 	updateParameterCache(false);
 
-	return get_landed_state();
+	if (get_freefall_state()) {
+		_state = LANDDETECTION_RES_FREEFALL;
+
+	} else if (get_landed_state()) {
+		_state = LANDDETECTION_RES_LANDED;
+
+	} else {
+		_state = LANDDETECTION_RES_FLYING;
+	}
+
+	return _state;
+}
+
+bool MulticopterLandDetector::get_freefall_state()
+{
+	if (_params.acc_threshold_m_s2 < 0.1f
+	    || _params.acc_threshold_m_s2 > 10.0f) {	//if parameter is set to zero or invalid, disable free-fall detection.
+		return false;
+	}
+
+	const uint64_t now = hrt_absolute_time();
+
+	float acc_norm = 0.0;
+
+	for (int i = 0; i < 3; i++) {
+		acc_norm += _sensors.accelerometer_m_s2[i] * _sensors.accelerometer_m_s2[i];
```

Comment:
@rousselmanu We should use the accelerations provided by the control state message. The estimator generates the control state message and it can e.g. switch sensor if one falls out. There is a PR which enabled this for ekf2 which I want to bring in this afternoon.